### ISSUE TITLE
Alternator: deleting a table leaves behind and unusable snapshot

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#5283]
By default, Scylla is configured to leave behind a snapshot (backup) of a deleted table.
This behavior will probably be unexpected by users of the DynamoDB API, which (as far as I can tell) doesn't do an automatic on-demand backup on delete) and can cause Alternator users to be surprised when deleting a table doesn't save them any space.

Moreover, not only are we doing this snapshot, nor did we implement the restore operation of the DynamoDB API (DynamoDB does allow restoring a deleted table from backup), so the resulting backup is not even usable.

We should either disable the snapshot-on-delete feature for Alternator tables, or support DynamoDB's API to restore them (and document somewhere that we do these automatic backups on delete, even though DynamoDB doesn't).

**Comment by nyh** [comment]
These unnecessary snapshots even happen when running alternator-test and all the tables get deleted at the end. Because we have a lot of tables with materialized views, we also get a lot of errors in the log about "Failed writing schema file" (this is #5659 or #5614). Even if we decide to leave snapshots enabled in Alternator (why?), we should definitely find a way to disable them when running Scylla for the test suite (my alternator-test/run script).

**Comment by nyh** [comment]
Some relevant code:

 `database::drop_column_family()` has a "snapshot" option, which defaults to true (the callers in `db/schema_tables.cc don't set it, so this default is used). This 'true' doesn't force the snapshot - it just means create a snapshot *if* the `auto_snapshot` option is one (see logic in `database::truncate()`). 
We can have `database::drop_column_family()` check if this is an Alternator table (by its keyspace name? any other way?) and if so, override the "snapshot" option to be false. This is ugly, but I'm not sure what else to suggest except drop this issue and leave the snapshots...

**Comment by nyh** [comment]
It turns out (see https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/pointintimerecovery_beforeyoubegin.html) that in Amazon DynamoDB, if a table has point-in-time-recovery (PITR) enabled and you delete the entire table, it's actually possible to restore this table using the usual restore syntax:
> If you delete a table with point-in-time recovery enabled, a system backup is automatically created and is retained for 35 days (at no additional cost). System backups allow you to restore the deleted table to the state it was in just before the point of deletion. All system backups follow a standard naming convention: `table-name$DeletedTableBackup`
We can emulate this convention and allow restoring `table-name$DeletedTableBackup` regardless of whether PITR (which we don't support yet) is enabled, as long as auto_snapshot is set. However, we'll also need some mechanism to remove old snapshots - assuming that DynamoDB applications don't use `nodetool`.

**Comment by nyh** [comment]
I think the better option for now would be to disable snapshots for Alternator tables - even if they are enabled by default for CQL for historic reasons. Even if we want to allow recovering of a deleted table, we should probably not do it with snapshots - we should do it by keeping and renaming the table - #7123 explains why snapshots are not a safe way to keep deleted tables alive.

**Comment by nyh** [comment]
I've upgraded this issue to "bug" - it can cause big problems for users not aware of this. If we want to keep these snapshots and remove this "bug" tag, we need to at least document this "feature" in compatibility.md.

**Comment by nyh** [comment]
> @nyh IMHO we could start by disabling snapshots creation altogether

Disable it by default, not just for Alternator?
I won't disagree (I explained in https://github.com/scylladb/scylladb/issues/7123 why I think the current "snapshot" feature is broken), but this needs to be decided not just by the Alternator developers, so if this is what you believe I suggest you try to suggest this in the mailing list (or as a new issue) and see what others think.

> and then work towards fixing them. Once the changes are ready and tested, we can enable snapshots again.

Fixing how? Snapshots are what they are. I suggested in https://github.com/scylladb/scylladb/issues/7123 to do something else, but that's not the "snapshots" we currently have.


> Currently it's the `database::drop_table_on_all_shards` method that has this `bool with_snapshot` parameter and it's the `truncate_table_on_all_shards` method that actually performs the check against `auto_snapshot` configuration. If we wanted to position a check for alternator, this would be the place: https://github.com/scylladb/scylladb/blob/master/replica/database.cc#L2382, but a better option would be to explicitly set `auto_snapshot: false` in `scylla.yaml` config for Alternator (if that's doable).

What do you mean "set ... in scylla.yaml config for Alternator"? We don't have a separate scylla.yaml for Alternator, and in general our default there is true :-(

If we **don't** do what I suggested in this issue (override auto-snapshot for Alternator tables), I think we should at least add a paragraph to alternator.md suggesting to set it to false. In fact, I'll send such a patch today, while we continue to debate this issue for three more years ;-)

**Comment by nyh** [comment]
Until we do something about this issue - one way or another - I added in https://github.com/scylladb/scylladb/pull/13103 a paragraph to docs/alternator/alternator.md, recommending that the user should disable this option manually. If we ever to solve *this* issue, we will need to remove that paragraph from the document.

**Comment by ptrsmrn** [comment]
> > @nyh IMHO we could start by disabling snapshots creation altogether
> 
> Disable it by default, not just for Alternator? I won't disagree (I explained in #7123 why I think the current "snapshot" feature is broken), but this needs to be decided not just by the Alternator developers, so if this is what you believe I suggest you try to suggest this in the mailing list (or as a new issue) and see what others think.

Yes, I meant disabling it for everyone/everywhere, since snapshots are not usable, but now I think that perhaps someone is using them either way, I/we just don't know about it/don't know how they are used, so it'd be safer to disable it just for the Alternator, as this is something we actually control and are (more) aware how it is used. 

> 
> > and then work towards fixing them. Once the changes are ready and tested, we can enable snapshots again.
> 
> Fixing how? Snapshots are what they are. I suggested in #7123 to do something else, but that's not the "snapshots" we currently have.

In general by "fixing" I meant providing a functionality to do something about them, possibly restore/delete them. 

> 
> > Currently it's the `database::drop_table_on_all_shards` method that has this `bool with_snapshot` parameter and it's the `truncate_table_on_all_shards` method that actually performs the check against `auto_snapshot` configuration. If we wanted to position a check for alternator, this would be the place: https://github.com/scylladb/scylladb/blob/master/replica/database.cc#L2382, but a better option would be to explicitly set `auto_snapshot: false` in `scylla.yaml` config for Alternator (if that's doable).
> 
> What do you mean "set ... in scylla.yaml config for Alternator"? We don't have a separate scylla.yaml for Alternator, and in general our default there is true :-(
> 
> If we **don't** do what I suggested in this issue (override auto-snapshot for Alternator tables), I think we should at least add a paragraph to alternator.md suggesting to set it to false. In fact, I'll send such a patch today, while we continue to debate this issue for three more years ;-)

OK, so setting `auto_snaphot` to `false` in the `scylla.yaml` is not an option. The other option is to disable it in the source code, possibly in the boolean expression I mentioned.

**Comment by mykaul** [comment]
> Yes, I meant disabling it for everyone/everywhere, since snapshots are not usable, but now I think that perhaps someone is using them either way, I/we just don't know about it/don't know how they are used, so it'd be safer to disable it just for the Alternator, as this is something we actually control and are (more) aware how it is used.

Sounds a bit harsh disable to all.

**Comment by ptrsmrn** [comment]
> > Yes, I meant disabling it for everyone/everywhere, since snapshots are not usable, but now I think that perhaps someone is using them either way, I/we just don't know about it/don't know how they are used, so it'd be safer to disable it just for the Alternator, as this is something we actually control and are (more) aware how it is used.
> 
> Sounds a bit harsh disable to all.

Sounds like it, but before dropping this idea, I'd maybe first consult @tzach if CQL users really use snaphots? Do we have some metrics/statistics on snapshot usage? Perhaps someone from support could help us better understand if disabling snapshots altogether by default (everywhere for everyone) is an option? (Because currently they're enabled by default; although this flag is commented out in [scylla.yaml](https://github.com/scylladb/scylladb/blob/master/conf/scylla.yaml#L366), we're enabling it by default in the source code [here](https://github.com/scylladb/scylladb/blob/master/db/config.cc#L526)).
Clients that want snapshots could still enable them themselves, so it's not like we are removing this option, just changing the default behavior. And btw. @nyh changing the deafult won't make the problem go away, Alternator would still create unusable snapshots if this option was turned on  :'-)
